lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+          # TODO: Implement a logging class to handle this.
+          # warn DEMUX_WARNING
+          @event_deque.each do |event|
+            event.result = response
+            event.set
+          end
+        else
+          start = 0
+          @in_deque.zip(@event_deque).each do |i, event|
+            response_copy = response.dup
+            subresponses = all_subresponses[start, i.count]
+            response_copy.send(@subresponse_field).clear
+            response_copy.send(@subresponse_field).concat(subresponses)
+            start += i.count
+            event.result = response_copy
+            event.set
+          end
+        end
+      rescue Exception => err
+        @event_deque.each do |event|
+          event.result = err
+          event.set
+        end
+      ensure
+        @in_deque.clear
+        @event_deque.clear
+      end
+
+      # This adds elements to the tasks.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @return [Event] an Event that can be used to wait on the response.
+      def extend(elts)
+        elts = [*elts]
+        @in_deque.push(elts)
+        event = event_for(elts)
+        @event_deque.push(event)
+        event
+      end
+
+      # Creates an Event that is set when the bundle with elts is sent.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @return [Event] an Event that can be used to wait on the response.
+      def event_for(elts)
+        event = Event.new
+        event.canceller = canceller_for(elts, event)
+        event
+      end
+
+      # Creates a cancellation proc that removes elts.
+      #
+      # The returned proc returns true if all elements were successfully removed
+      # from @in_deque and  @event_deque.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @param [Event] an Event that can be used to wait on the response.
+      # @return [Proc] the canceller that when called removes the elts
+      #     and events.
+      def canceller_for(elts, event)
+        proc do
+          event_index = @event_deque.find_index(event) || -1
+          in_index = @in_deque.find_index(elts) || -1
+          @event_deque.delete_at(event_index) unless event_index == -1
+          @in_deque.delete_at(in_index) unless in_index == -1
+          if event_index == -1 || in_index == -1
+            false
+          else
+            true
+          end
+        end
+      end
+
+      # private_class_method :run_with_no_subresponse,
+      #                      :run_with_subresponses,
+      #                      :event_for, :cancellor_for
+    end
+
+    # Organizes bundling for an api service that requires it.
+    class Executor
+      # @param [BundleOptions]configures strategy this instance
```

Comment:
nit: options is missing (`@param options [BundleOptions] configures strategy`...)